docs/spokeclusterhealthcheck.md

```diff
@@ -0,0 +1,60 @@
+# Spoke cluster available check
+
+This document describes a mechanism on how hub cluster check the availability of agent running
+on spoke cluster, a new condition `SpokeClusterConditionAvailable` will be introduced to describe
+whether a spoke cluster is available
+
+If the kube-apiserver is health and the registration/work agent is running with the minimum
+deployment on a spoke cluster, this spoke cluster will be considered available
+
+The Kubernetes Metrics API may be used to get the status of spoke cluster kube-apiserver or spoke
+registration/work agent, but Metrics API has some limits for our cases
+
+- The hub may not access the spoke cluster, so we cannot use Kubernetes Metrics API to
+get spoke cluster metrics from hub cluster
+- Reporting metrics will require additional request path from hub to spoke (or spoke to hub)
+- To know whether a spoke cluster is available, the hub may need periodically poll all of spoke
+clusters, this is an inefficient way
+
+So we propose hub and spoke coordinate a condition to present whether a spoke cluster is available
+
+Actors
+
+- agent on spoke cluster
+- hub controller on hub cluster
+
+## Available check
+
+1. after a spoke cluster is accepted by hub cluster admin, the agent on spoke cluster will
+create a [`Lease`](https://github.com/kubernetes/api/blob/master/coordination/v1/types.go#L27)
+on its namespace on hub cluster, then the agent will periodically (by default 10s, the period time
+should be adjusted, see the scale problem section)
+    - update the `RenewTime` of `Lease` to keep available, and
+    - check the spoke cluster kube-apiserver health by `healthz` API, and check the
+    registration/work agent is running with the minimum deployment, if all of them are
+    available the agent make sure the status of `SpokeClusterConditionAvailable`
+    condition is `true`, otherwise, the agent updates the status to `false`
+2. hub controller lists the spoke clusters on hub cluster periodically (by default 5s), for
```

Comment:
an informer ends up taking many watch events, but the smearing may help.  I guess I would at least try an informer to start because it's "normal".

You probably want to label these leases to avoid selecting leases for nodes.